### PR TITLE
use libcurl

### PR DESCRIPTION
**Pull request by cangfengzhs**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [ ] enhancement

#### Description:
Use libcurl instead of invoke "/usr/bin/curl" with popen

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/codec/test/CMakeLists.txt [hunk 1]
```diff
@@ -54,6 +54,7 @@ nebula_add_test(
         ${PROXYGEN_LIBRARIES}
         wangle
         gtest
+        curl
```
[the remark this answers is not in the document]
Comment (cangfengzhs) [reply src/codec/test/CMakeLists.txt:57]: I also have this puzzle. However, if I do not add curls, these objects will report errors when linking. At the same time, I'm worried that the addition of curl will slow down compilation

[the remark this answers is not in the document]
Comment (cangfengzhs) [reply src/codec/test/CMakeLists.txt]: I checked and found that the reason is that we added many unnecessary dependent objects to the target of many execution files. If they include http_client, must link curl. I have deleted some.

[the remark this answers is not in the document]
Comment (panda-sheep) [reply src/codec/test/CMakeLists.txt]: 👍

File: src/common/http/test/CMakeLists.txt [hunk 2]
```diff
@@ -22,5 +23,6 @@ nebula_add_test(
     LIBRARIES
         ${PROXYGEN_LIBRARIES}
         gtest
-        gtest_main
-)
+        curl
+
+)
```
Comment (panda-sheep) [inline src/common/http/test/CMakeLists.txt:28]: Add a newline?

File: src/common/http/test/CMakeLists.txt [hunk 3]
```diff
@@ -8,6 +8,7 @@ nebula_add_test(
     SOURCES
         HttpClientTest.cpp
     OBJECTS
+        $<TARGET_OBJECTS:base_obj>
```
Comment (panda-sheep) [inline src/common/http/test/CMakeLists.txt:11]: Line 19 already exists?

File: src/meta/CMakeLists.txt [hunk 4]
```diff
@@ -139,7 +139,7 @@ set(meta_test_deps
     $<TARGET_OBJECTS:file_based_cluster_id_man_obj>
     $<TARGET_OBJECTS:ws_obj>
     $<TARGET_OBJECTS:ws_common_obj>
-    $<TARGET_OBJECTS:http_client_obj>
+    # $<TARGET_OBJECTS:http_client_obj>
```
Comment (panda-sheep) [inline src/meta/CMakeLists.txt:142]: can remove this comment？

**Review by panda-sheep — COMMENTED** [review]
Good job~

File: src/common/http/test/CMakeLists.txt (hunk 2 above)
Comment (cangfengzhs) [reply src/common/http/test/CMakeLists.txt]: Done

File: src/common/http/test/CMakeLists.txt (hunk 3 above)
Comment (cangfengzhs) [reply src/common/http/test/CMakeLists.txt:11]: Done

File: src/meta/CMakeLists.txt (hunk 4 above)
Comment (cangfengzhs) [reply src/meta/CMakeLists.txt]: Done

File: src/common/http/HttpClient.h [hunk 5]
```diff
@@ -1,47 +1,63 @@
...
+ public:
+  static CurlHandle* instance();
+
+ private:
+  CurlHandle();
+  ~CurlHandle();
+};
+
+struct HttpResponse {
+  CURLcode curlCode;
```
Comment (critical27) [inline src/common/http/HttpClient.h:27]: `CURLcode` is a int not a enum? Didn't find it in this PR

File: src/common/http/HttpClient.h [hunk 6]
```diff
@@ -1,47 +1,63 @@
...
-  static StatusOr<std::string> sendRequest(const std::string& path,
-                                           const folly::dynamic& data,
-                                           const std::string& reqType);
+  static HttpResponse get(const std::string& url);
+  static HttpResponse get(const std::string& url, const std::vector<std::string>& headers);
+
+  static HttpResponse post(const std::string& url,
+                           const std::vector<std::string>& headers,
+                           const std::string& body);
+  static HttpResponse delete_(const std::string& url, const std::vector<std::string>& headers);
```
Comment (critical27) [inline src/common/http/HttpClient.h:41]: `del`? Not a big issue, up to you

**Review by critical27 — APPROVED** [review]
Good job

File: src/common/http/HttpClient.h (hunk 5 above)
Comment (cangfengzhs) [reply src/common/http/HttpClient.h:27]: It's an integer and defined by libcurl

File: src/common/http/HttpClient.h (hunk 6 above)
Comment (cangfengzhs) [reply src/common/http/HttpClient.h:41]: Emm, both are OK. But I don't want to change it:)
Maybe someone who doesn't like delete_ will modify it one day.

**Review by panda-sheep — APPROVED** [review]

**Review by yixinglu — COMMENTED** [review]
Is the libcurl included in nebula third parties?  or use  the local system library?

**Comment by cangfengzhs** [comment]
> Is the libcurl included in nebula third parties? or use the local system library?

https://github.com/vesoft-inc/nebula-third-party



> Is the libcurl included in nebula third parties? or use the local system library?

It's included in third-party

**Pull request merged**